### ISSUE TITLE
dotnet 8.0.200 breaking change - wrong exception is thrown when rethrow in a catch inside another catch using async methods

### ISSUE THREAD

**Issue by yp87** [dotnet/roslyn#72141]
### Describe the bug
Upgrading to dotnet 8.0.200 (repro with 8.0.201) caused a breaking change when rethrowing in a catch within a catch. The first exception is rethrown instead of the last exception.

### To Reproduce
Using sync methods, or throwing directly, would not reproduce the issue.
```
using System.Threading.Tasks;

Task ThrowArgumentNullExceptionAsync()
{
    throw new ArgumentNullException();
    return Task.CompletedTask;
}

Task ThrowInvalidOperationExceptionAsync()
{
    throw new InvalidOperationException();
    return Task.CompletedTask;
}

try
{
    // throw new ArgumentNullException(); <--- also reproduce
    await ThrowArgumentNullExceptionAsync();
}
catch (ArgumentNullException)
{
    try
    {
        await ThrowInvalidOperationExceptionAsync();
    }
    catch (InvalidOperationException)
    {
        throw;
    }
}
```

### Exceptions (if any)
This program throws ArgumentNullException instead of InvalidOperationException like in previous versions (8.0.101).

### Further technical details
.NET SDK:
 Version:           8.0.200
 Commit:            438cab6a9d
 Workload version:  8.0.200-manifests.e575128c

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.200\

.NET workloads installed:
 [wasm-tools]
   Installation Source: VS 17.9.34607.119
   Manifest Version:    8.0.2/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.workload.mono.toolchain.current\8.0.2\WorkloadManifest.json
   Install Type:        FileBased

Host:
  Version:      8.0.2
  Architecture: x64
  Commit:       1381d5ebd2

.NET SDKs installed:
  6.0.100-preview.7.21379.14 [C:\Program Files\dotnet\sdk]
  8.0.200 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.0-preview.7.21378.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.27 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.0-preview.7.21377.19 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-rc.1.21451.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.20 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.26 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.27 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.0-preview.7.21378.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-rc.1.21451.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.26 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.27 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.16 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  arm64 [C:\Program Files\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\arm64\InstallLocation]
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download

**Comment by cston** [comment]
@jjonescz, is this a duplicate of https://github.com/dotnet/roslyn/issues/71569?

**Comment by jjonescz** [comment]
Yes, it's a duplicate.

**Comment by karaziox** [comment]
Any chance of a servicing release for this? We've seen this actively altering behavior in production code. I'm confused how this bug was known and still made it through to an official release, that's a pretty nasty miscompilation

**Comment by cston** [comment]
cc @jcouv, @jaredpar

**Comment by jjonescz** [comment]
FYI: fix for this has been released in SDK 8.0.203 and VS 17.9.2
